pkg/policy/selectorcache.go

```diff
@@ -223,12 +239,46 @@ func (sc *SelectorCache) GetModel() models.SelectorCache {
 	return selCacheMdl
 }
 
+func (sc *SelectorCache) handleUserNotifications() {
+	for {
+		sc.userMutex.Lock()
+		for len(sc.userNotes) == 0 {
+			sc.userCond.Wait()
+		}
+		// get the current batch of notifications and release the lock so that SelectorCache
+		// can't block on userMutex while we call IdentitySelectionUpdated callbacks below.
+		notifications := sc.userNotes
+		sc.userNotes = nil
+		sc.userMutex.Unlock()
+
+		for _, n := range notifications {
+			n.user.IdentitySelectionUpdated(n.selector, n.added, n.deleted)
+		}
+	}
+}
+
+func (sc *SelectorCache) queueUserNotification(user CachedSelectionUser, selector CachedSelector, added, deleted []identity.NumericIdentity) {
+	sc.userMutex.Lock()
```

Comment:
[1] Replace the mutex with a 2-slot semaphore. Then call `sc.userCond.Signal()` until we are able to acquire `sc.semaphore.Acquire(2)` in this line, which means that `handleUserNotifications` will only re-acquire the semaphore until this function has released it because `handleUserNotifications` will only do a `sc.semaphore.Acquire(1)`.

Reply:
IMO this is not necessary and would be more complex.